apps/src/netsim/NetSimApiError.js

```diff
@@ -0,0 +1,68 @@
+/**
+ * @overview Extended error type returned for failed interactions with NetSim
+ * server API (net_sim_api.rb) that extracts and exposes additional error info.
+ */
+/* jshint
+ funcscope: true,
+ newcap: true,
+ nonew: true,
+ shadow: false,
+ unused: true,
+
+ maxlen: 90,
+ maxparams: 3,
+ maxstatements: 200
+ */
+'use strict';
+
+require('../utils'); // provide Function.prototype.inherits
+
+/**
+ * Special error type for failed server requests, which tries to extract
+ * additional error information from the server's response.
+ * @param {jqXHR} request
+ * @constructor
+ * @extends Error
+ */
+var NetSimApiError = module.exports = function (request) {
+  /** @type {string} */
+  this.name = 'NetSimApiError';
+
+  /** @type {string} */
+  this.message = 'Request failed';
+
+  /** @type {string} */
+  this.stack = (new Error()).stack;
+
+  /**
+   * Additional error information returned by the server, which can drive
+   * specific responses by the client.
+   * @type {string|Array}
+   */
+  this.details = undefined;
+
+  // Attempt to extract additional information from the request object
+  if (request) {
+    this.message = 'status: ' + request.status + '; error: ' + request.statusText;
+    try {
+      var responseJson = JSON.parse(request.responseText);
+      if (responseJson.details) {
+        this.details = responseJson.details;
+        this.message += '; details: ' + JSON.stringify(this.details);
+      }
+    } catch (e) {
```

Comment:
If you are constructing `NetSimApiError` then presumably the request has failed either way, it's just a question of whether we can extract additional error details.

Most of our stock responses (defined in core.rb) including `no_content` and `not_authorized` which are still used by NetSim, don't have a JSON body and would throw an exception in `JSON.parse`.  In this case I'd like to catch that exception and harmlessly finish constructing the error with an empty `details` field.  Only the `json_bad_request` response has a parse-able JSON body.